File: simulators/typescript/src/scenarios/basicCrud.ts
import { Pool } from "pg";
import { Config, RawResult } from "../types";

export async function basicCrud(config: Config): Promise<RawResult> {
  const pool = new Pool({ host: config.host, port: config.port, user: config.user, password: config.password, database: config.database, max: config.concurrency });
  pool.on("error", () => {}); // suppress unhandled pool-level errors
  let ops = 0, errors = 0;
  const latencies: number[] = [];

  const start = Date.now();
  const workers = Array.from({ length: config.concurrency }, (_, i) => (async () => {
    for (let iter = 0; iter < 20; iter++) {
      const email = `crud_${i}_${iter}@test.com`;

      // INSERT
      let t = performance.now();
      let userId: number;
      try {
        const res = await pool.query("INSERT INTO users (name, email, balance) VALUES ($1, $2, $3) RETURNING id", [`crud_user_${i}_${iter}`, email, 100.0]);
        userId = res.rows[0].id;
      } catch { errors++; ops++; latencies.push(performance.now() - t); continue; }
      latencies.push(performance.now() - t); ops++;

      // SELECT
      t = performance.now();
      try { await pool.query("SELECT name FROM users WHERE id = $1", [userId]); }
      catch { errors++; }
      latencies.push(performance.now() - t); ops++;

      // UPDATE
      t = performance.now();
      try { await pool.query("UPDATE users SET balance = balance + 50 WHERE id = $1", [userId]); }
      catch { errors++; }
      latencies.push(performance.now() - t); ops++;

      // DELETE
      t = performance.now();
      try { await pool.query("DELETE FROM users WHERE id = $1", [userId]); }
      catch { errors++; }
      latencies.push(performance.now() - t); ops++;
    }
  })());

  await Promise.all(workers);
  await pool.end();
  return { totalOps: ops, errors, durationMs: Date.now() - start, latencies };
}

File: simulators/go/scenarios/basic_crud.go
package scenarios

import (
	"context"
	"fmt"
	"sync"
	"sync/atomic"
	"time"
)

// BasicCRUD exercises INSERT, SELECT, UPDATE, DELETE operations concurrently.
func BasicCRUD(ctx context.Context, cfg *Config) (*Result, error) {
	var (
		ops       atomic.Int64
		errCount  atomic.Int64
		mu        sync.Mutex
		latencies []time.Duration
	)
	errs := NewErrorSampler(10)

	start := time.Now()
	var wg sync.WaitGroup

	for i := 0; i < cfg.Concurrency; i++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()

			conn, err := ConnectClient(ctx, cfg)
			if err != nil {
				ops.Add(1)
				errCount.Add(1)
				errs.Record(err)
				return
			}
			defer conn.Close(ctx)

			for iter := 0; iter < 20; iter++ {
				email := fmt.Sprintf("crud_%d_%d@test.com", workerID, iter)

				// INSERT
				t := time.Now()
				var userID int
				err := conn.QueryRow(ctx, "INSERT INTO users (name, email, balance) VALUES ($1, $2, $3) RETURNING id",
					fmt.Sprintf("crud_user_%d_%d", workerID, iter), email, 100.00).Scan(&userID)
				d := time.Since(t)
				mu.Lock()
				latencies = append(latencies, d)
				mu.Unlock()
				ops.Add(1)
				if err != nil {
					errCount.Add(1)
					errs.Record(err)
					continue
				}

				// SELECT
				t = time.Now()
				var name string
				err = conn.QueryRow(ctx, "SELECT name FROM users WHERE id = $1", userID).Scan(&name)
				d = time.Since(t)
				mu.Lock()
				latencies = append(latencies, d)
				mu.Unlock()
				ops.Add(1)
				if err != nil {
					errCount.Add(1)
					errs.Record(err)
					continue
				}

				// UPDATE
				t = time.Now()
				_, err = conn.Exec(ctx, "UPDATE users SET balance = balance + 50 WHERE id = $1", userID)
				d = time.Since(t)
				mu.Lock()
				latencies = append(latencies, d)
				mu.Unlock()
				ops.Add(1)
				if err != nil {
					errCount.Add(1)
					errs.Record(err)
					continue
				}

				// DELETE
				t = time.Now()
				_, err = conn.Exec(ctx, "DELETE FROM users WHERE id = $1", userID)
				d = time.Since(t)
				mu.Lock()
				latencies = append(latencies, d)
				mu.Unlock()
				ops.Add(1)
				if err != nil {
					errCount.Add(1)
					errs.Record(err)
				}
			}
		}(i)
	}

	wg.Wait()
	duration := time.Since(start)

	return &Result{
		TotalOps:  int(ops.Load()),
		Errors:    int(errCount.Load()),
		Duration:  duration,
		Latencies: latencies,
		Notes:     errs.Notes(),
	}, nil
}
